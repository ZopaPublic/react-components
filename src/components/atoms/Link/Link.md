### Summary

Use `<Link />` to create hyperlinks to other web pages, files, locations within the same page, email addresses, or any other URL.

⚠️ &nbsp; Links are never headings

### Examples

- Normal link

```tsx
import { Text, Link } from '@zopauk/react-components';

<Text as="p">
  Some text with
  <Link href="http://duckduckgo.com" onClick={() => alert('Link clicked!')}>
    a link
  </Link>
</Text>;
```

- Button link

```tsx
import { Text, Link } from '@zopauk/react-components';

<Text as="p">
  Some text with
  <Link as="button" href="http://duckduckgo.com" onClick={() => alert('Link clicked!')}>
    a link
  </Link>
</Text>;
```

- Negative link

```tsx { "props": { "style": { "backgroundColor": "#00B9A7", "border": "none" } } }
import { Text, Link, colors } from '@zopauk/react-components';

<Text as="p" color={colors.greyDarkest}>
  Some text with
  <Link href="http://duckduckgo.com" onClick={() => alert('Link clicked!')} negative>
    a link
  </Link>
</Text>;
```

- With `target="_blank"` ( _notice a square icon appeared_ )

```tsx
import { Text, Link } from '@zopauk/react-components';

<Text as="p">
  Some text with
  <Link target="_blank" href="http://duckduckgo.com" onClick={() => alert('Link clicked!')}>
    a link
  </Link>
</Text>;
```

- Link as a custom component

```tsx
import React from 'react';
import styled from 'styled-components';
import { linkStyle } from '@zopauk/react-components';

// could be a gatsby or react-router-dom <Link />
const SomeLibraryLink = ({ href, ...rest }) => <a href={href} {...rest} />;

const CustomLink = styled(SomeLibraryLink)`
  ${linkStyle}
`;

<CustomLink href="https://www.zopa.com">Custom link</CustomLink>;
```
