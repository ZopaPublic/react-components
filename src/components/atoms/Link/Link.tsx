import React from 'react';
import styled, { css } from 'styled-components';
import { colors, typography } from '../../../constants';
import { useThemeContext } from '../../styles/Theme';
export interface LinkProps
  extends React.AnchorHTMLAttributes<HTMLAnchorElement>,
    React.RefAttributes<HTMLAnchorElement> {
  target?: '_blank';
  negative?: boolean;
  as?: 'a' | 'button';
  showTargetIcon?: boolean;
}

export const linkStyle = css<LinkProps>`
  background-color: transparent;
  font-size: inherit;
  font-family: ${typography.primary};
  font-weight: ${typography.weights.semiBold};
  line-height: inherit;
  color: ${({ negative }) => (negative ? colors.greyDarkest : colors.actionPlain)};
  cursor: pointer;
  text-decoration: underline;
  user-select: auto;
  appearance: none;
  padding: 0;
  border: none;

  &:hover,
  &:active {
    color: ${({ negative }) => (negative ? 'black' : colors.actionDark)};

    svg path {
      fill: ${({ negative }) => (negative ? 'black' : colors.actionDark)};
    }
  }
`;

const StyledLink = styled.a<LinkProps>`
  ${linkStyle};
  font-family: ${({ theme }) => theme.typography.primary};
  font-weight: ${({ theme }) => theme.link.weight};
  color: ${({ theme }) => theme.link.color};
  &:hover {
    color: ${({ theme }) => theme.link.hover.color};
    svg path {
      fill: ${({ negative, theme }) => (negative ? 'black' : theme.link.hover.color)};
    }
  }
  &:active {
    color: ${({ theme }) => theme.link.active.color};
    svg path {
      fill: ${({ negative, theme }) => (negative ? 'black' : theme.link.active.color)};
    }
  }
`;

const TargetIconWrapper = styled.svg`
  margin-left: 6px;
  top: 2px;
  position: relative;
`;

const TargetIcon = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <TargetIconWrapper xmlns="http://www.w3.org/2000/svg" width="15" height="15">
      <g fill={props.color || colors.actionPlain}>
        <path d="M2.6 4H0v11h11v-2.6H2.6z" />
        <path d="M4 0v11h11V0H4zm9.7 1.3v8.4L5.2 1.3h8.5z" />
      </g>
    </TargetIconWrapper>
  );
};

const Link = React.forwardRef<HTMLAnchorElement, LinkProps>(
  ({ children, color = colors.actionPlain, target, showTargetIcon = true, ...rest }, ref) => {
    const theme = useThemeContext();

    return (
      <StyledLink ref={ref} color={color} target={target} {...rest} theme={theme}>
        {children}
        {target === '_blank' && showTargetIcon && !theme.link.disableTargetIcon && <TargetIcon color={color} />}
      </StyledLink>
    );
  },
);

Link.defaultProps = {
  color: colors.actionPlain,
};

export default Link;
