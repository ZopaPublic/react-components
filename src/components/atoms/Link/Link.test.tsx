import axe from '../../../../axe-helper';
import React from 'react';
import { render } from '@testing-library/react';
import Link from './Link';

describe('<Link />', () => {
  it('renders the link with default values and the svg is hidden with no a11y violations', async () => {
    const { container } = render(<Link href="http://duckduckgo.com">text</Link>);
    expect(container.firstChild).toMatchSnapshot();

    const results = await axe(container.innerHTML);
    expect(results).toHaveNoViolations();
  });

  it('renders the link with target _blank with the svg visible', () => {
    const { container } = render(
      <Link href="http://duckduckgo.com" target="_blank">
        text
      </Link>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('doesnt render the link with a target icon', () => {
    const { container } = render(
      <Link href="http://duckduckgo.com" showTargetIcon={false} target="_blank">
        text
      </Link>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders for being displayed on top of dark backgrounds', () => {
    const { container } = render(
      <Link href="http://duckduckgo.com" target="_blank" negative>
        text
      </Link>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
