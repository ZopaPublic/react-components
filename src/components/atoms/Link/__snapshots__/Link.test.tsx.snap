// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> doesnt render the link with a target icon 1`] = `
.c0 {
  background-color: transparent;
  font-size: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  line-height: inherit;
  color: #3B46C4;
  cursor: pointer;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-user-select: auto;
  -moz-user-select: auto;
  -ms-user-select: auto;
  user-select: auto;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 0;
  border: none;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  color: auto;
}

.c0:hover,
.c0:active {
  color: #21247F;
}

.c0:hover svg path,
.c0:active svg path {
  fill: #21247F;
}

.c0:hover {
  color: auto;
}

.c0:hover svg path {
  fill: auto;
}

.c0:active {
  color: auto;
}

.c0:active svg path {
  fill: auto;
}

<a
  class="c0"
  color="#3B46C4"
  href="http://duckduckgo.com"
  target="_blank"
>
  text
</a>
`;

exports[`<Link /> renders for being displayed on top of dark backgrounds 1`] = `
.c0 {
  background-color: transparent;
  font-size: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  line-height: inherit;
  color: #2C3236;
  cursor: pointer;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-user-select: auto;
  -moz-user-select: auto;
  -ms-user-select: auto;
  user-select: auto;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 0;
  border: none;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  color: auto;
}

.c0:hover,
.c0:active {
  color: black;
}

.c0:hover svg path,
.c0:active svg path {
  fill: black;
}

.c0:hover {
  color: auto;
}

.c0:hover svg path {
  fill: black;
}

.c0:active {
  color: auto;
}

.c0:active svg path {
  fill: black;
}

.c1 {
  margin-left: 6px;
  top: 2px;
  position: relative;
}

<a
  class="c0"
  color="#3B46C4"
  href="http://duckduckgo.com"
  target="_blank"
>
  text
  <svg
    class="c1"
    height="15"
    width="15"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      fill="#3B46C4"
    >
      <path
        d="M2.6 4H0v11h11v-2.6H2.6z"
      />
      <path
        d="M4 0v11h11V0H4zm9.7 1.3v8.4L5.2 1.3h8.5z"
      />
    </g>
  </svg>
</a>
`;

exports[`<Link /> renders the link with default values and the svg is hidden with no a11y violations 1`] = `
.c0 {
  background-color: transparent;
  font-size: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  line-height: inherit;
  color: #3B46C4;
  cursor: pointer;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-user-select: auto;
  -moz-user-select: auto;
  -ms-user-select: auto;
  user-select: auto;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 0;
  border: none;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  color: auto;
}

.c0:hover,
.c0:active {
  color: #21247F;
}

.c0:hover svg path,
.c0:active svg path {
  fill: #21247F;
}

.c0:hover {
  color: auto;
}

.c0:hover svg path {
  fill: auto;
}

.c0:active {
  color: auto;
}

.c0:active svg path {
  fill: auto;
}

<a
  class="c0"
  color="#3B46C4"
  href="http://duckduckgo.com"
>
  text
</a>
`;

exports[`<Link /> renders the link with target _blank with the svg visible 1`] = `
.c0 {
  background-color: transparent;
  font-size: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  line-height: inherit;
  color: #3B46C4;
  cursor: pointer;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-user-select: auto;
  -moz-user-select: auto;
  -ms-user-select: auto;
  user-select: auto;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 0;
  border: none;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  color: auto;
}

.c0:hover,
.c0:active {
  color: #21247F;
}

.c0:hover svg path,
.c0:active svg path {
  fill: #21247F;
}

.c0:hover {
  color: auto;
}

.c0:hover svg path {
  fill: auto;
}

.c0:active {
  color: auto;
}

.c0:active svg path {
  fill: auto;
}

.c1 {
  margin-left: 6px;
  top: 2px;
  position: relative;
}

<a
  class="c0"
  color="#3B46C4"
  href="http://duckduckgo.com"
  target="_blank"
>
  text
  <svg
    class="c1"
    height="15"
    width="15"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      fill="#3B46C4"
    >
      <path
        d="M2.6 4H0v11h11v-2.6H2.6z"
      />
      <path
        d="M4 0v11h11V0H4zm9.7 1.3v8.4L5.2 1.3h8.5z"
      />
    </g>
  </svg>
</a>
`;
